### PR TITLE
Validation for the parameter enqueued in to_internal_repr of FloatDistribution and IntDistribution

### PR DESCRIPTION
**Pull request by cross32768**
Note: this PR depends on PR https://github.com/optuna/optuna/pull/4126. Please review and merge it first.

## Motivation

Resolve https://github.com/optuna/optuna/issues/4070

## Description of the changes

Add validation to raise error when value <= 0.0 is enqueued with log=True.
In addition, I add to error handling when not a number is enqueued.
I also added test cases to validate this change.

### CODE REVIEW

File: tests/test_distributions.py [hunk 1]
```diff
@@ -213,9 +213,21 @@ def test_int_internal_representation(value: Any) -> None:
...
+    "value, kwargs",
+    [
+        ("foo", {}),
+        ((), {}),
+        ([], {}),
+        ({}, {}),
+        (set(), {}),
+        (np.ones(2), {}),
+        (np.nan, {}),
+        (0, dict(log=True)),
```
Comment (knshnb) [inline tests/test_distributions.py:226]:
```suggestion
        (0, dict(log=True)),
        (-1, dict(log=True)),
```

File: tests/test_distributions.py [hunk 2]
```diff
@@ -234,9 +246,21 @@ def test_float_internal_representation(value: Any) -> None:
...
```
Comment (knshnb) [inline tests/test_distributions.py:259]:
```suggestion
        (0.0, dict(log=True)),
        (-1.0, dict(log=True)),
```

File: tests/test_distributions.py (hunk 1 above)
Comment (knshnb) [reply tests/test_distributions.py:226]: How about adding a test case of negative values?

File: tests/test_distributions.py (hunk 2 above)
Comment (knshnb) [reply tests/test_distributions.py:259]: Ditto.

File: optuna/trial/_trial.py [hunk 3]
```diff
@@ -597,6 +597,7 @@ def _suggest(self, name: str, distribution: BaseDistribution) -> Any:
                     study, trial, name, distribution
                 )
 
+            # param_value is validated here (invalid value like np.nan raises ValueError)
```
Comment (knshnb) [inline optuna/trial/_trial.py:600]:
```suggestion
            # `param_value` is validated here (invalid value like np.nan raises ValueError).
```
Please also refer to https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#optuna-specific-conventions.

**Review by knshnb — COMMENTED** [review]
Thanks for the PR. It looks good overall! Please take a look at some small comments I left.

**Review by HideakiImamura — DISMISSED** [review]
LGTM except for @knshnb's comments.

**Comment by cross32768** [comment]
Could someone check the updated document? (source: study.py)

File: optuna/study/study.py [hunk 4]
```diff
@@ -806,6 +806,14 @@ def objective(trial):
         Args:
             params:
                 Parameter values to pass your objective function.
+
+                .. note::
+                    This method raises ``ValueError`` for invalid parameter values
+                    like not a number (e.g. ``np.nan``),
+                    non-positive value when parameter is log-scale,
+                    value which cannot be casted to float for int/float distribution,
+                    value which does not be included in choices for categorical distribution, etc.
```
Comment (HideakiImamura) [inline optuna/study/study.py:815]: The error does not occur in the call of `enqueue_trial`. The error occurs in the next call of `suggest_***`. There is some discussions as to where such documentation should be written. I think it should be in `suggest_***` or `***Distribution`. (I prefer the later.)

By the way, we have [a rule to document the error](https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#exceptions).  I think we should NOT document the error when the parameter value is invalid, since it is obvious that such parameters are not acceptable. However, the obviousness is open for discussion.

Comment (cross32768) [reply optuna/study/study.py:815]: Thank you for your comment.
I agree to write such a kind of documentation in ``***Distribution``.

And sorry I had insufficient understanding for a rule to document the error.
I think error caused by not a number or values which cannot be casted to float/int is obvious, but non-positive value with log=True option and values does not be included in choices may require more discussion.

**Comment by cross32768** [comment]
For a moment, I remove the added document

File: optuna/study/study.py (hunk 4 above)
Comment (knshnb) [reply optuna/study/study.py]: I basically agree with the discussion. IMO, only it should be documented in the `log` argument of `***Distribution` that non-positive values are not allowed because this PR changes the behavior.

Comment (cross32768) [reply optuna/study/study.py]: I updated the code in response to this comment.

**Review by knshnb — APPROVED** [review]
Thanks for the update. LGTM.
